### PR TITLE
ARROW-13541: [C++][Python] Implement ExtensionScalar

### PR DESCRIPTION
**Pull request by pitrou**

### CODE REVIEW

File: python/pyarrow/scalar.pxi [hunk 1]
```diff
@@ -838,6 +843,67 @@ cdef class UnionScalar(Scalar):
...
+    Concrete class for Extension scalars.
+    """
+
+    @property
+    def value(self):
+        """
+        Return storage value as a scalar.
+        """
+        cdef CExtensionScalar* sp = <CExtensionScalar*> self.wrapped.get()
+        return Scalar.wrap(sp.value) if sp.is_valid else None
```
Comment (jorisvandenbossche) [inline python/pyarrow/scalar.pxi:857]: This could also be a null scalar of the storage type?

File: python/pyarrow/scalar.pxi [hunk 2]
```diff
@@ -838,6 +843,67 @@ cdef class UnionScalar(Scalar):
...
+
+    def as_py(self):
+        """
+        Return this scalar as a Python object.
+        """
+        # XXX should there be a hook to wrap the result in a custom class?
```
Comment (jorisvandenbossche) [inline python/pyarrow/scalar.pxi:863]: Similarly as we have `__arrow_ext_class__` for using an ExtensionArray subclass for the array object, we could indeed have a way to specify a custom scalar wrapper (but that's also fine to leave as a future enhancement I think)

**Review by jorisvandenbossche — COMMENTED** [review]
Looks good, thanks! One small comment. I also tested it with pandas' extension types, and I am wondering if it's the intention that casting extension scalars should now work:

```
In [35]: arr = pa.array(pd.period_range("2012", periods=3, freq="D"))

In [36]: arr
Out[36]: 
<pyarrow.lib.ExtensionArray object at 0x7f5ebbe38b20>
[
  15340,
  15341,
  15342
]

In [37]: arr.type
Out[37]: ArrowPeriodType(DataType(int64))

In [38]: arr.cast(pa.int64())
Out[38]: 
<pyarrow.lib.Int64Array object at 0x7f5ebbe307c0>
[
  15340,
  15341,
  15342
]

In [39]: arr[0].cast(pa.int64())
---------------------------------------------------------------------------
ArrowNotImplementedError                  Traceback (most recent call last)
<ipython-input-39-6681c4ea2967> in <module>
----> 1 arr[0].cast(pa.int64())

~/scipy/repos/arrow/python/pyarrow/scalar.pxi in pyarrow.lib.Scalar.cast()

~/scipy/repos/arrow/python/pyarrow/error.pxi in pyarrow.lib.pyarrow_internal_check_status()

~/scipy/repos/arrow/python/pyarrow/error.pxi in pyarrow.lib.check_status()

ArrowNotImplementedError: cast to int64 from extension<pandas.period<ArrowPeriodType>>
../src/arrow/scalar.cc:991  VisitTypeInline(*to, &unpack_to_type)
```

So casting the array works, but not the scalar.

**Comment by jorisvandenbossche** [comment]
(I see that in scalar.cc we still have a NotImplemented for extension types in CastImplVisitor, so my example was expected)

**Review by bkietz — APPROVED** [review]

File: python/pyarrow/scalar.pxi (hunk 1 above)
Comment (pitrou) [reply python/pyarrow/scalar.pxi:857]: Returning `None` mirrors the behaviour on the C++ side, though.

File: python/pyarrow/scalar.pxi (hunk 2 above)
Comment (pitrou) [reply python/pyarrow/scalar.pxi:863]: Feel free to open a JIRA, as you probably know the desired behaviour better than me :-)

**Comment by pitrou** [comment]
@jorisvandenbossche The cast kernels work... but the cast method on scalars is implemented separately for the moment. I believe there's a JIRA open to address that.

File: python/pyarrow/scalar.pxi (hunk 2 above)
Comment (jorisvandenbossche) [reply python/pyarrow/scalar.pxi:863]: Opened https://issues.apache.org/jira/browse/ARROW-13612

**Pull request closed without merging**